### PR TITLE
fix: updated dvp swap contract 

### PR DESCRIPTION
**Pull request by SaeeDawod**
made subgraphs and tests for the updated dvpswap contract.
Screenshot 2025-04-25 at 19 42 17

## Summary by Sourcery

Replace the existing DvPSwap contract with a new XvPSettlement contract, providing a more robust and flexible settlement mechanism for token exchanges

New Features:
- Introduced XvPSettlement contract with multi-flow token exchange support
- Added support for atomic settlements with approval and execution mechanisms
- Implemented deterministic contract deployment using CREATE2

Enhancements:
- Improved settlement flow with more granular control
- Added support for auto-execution of settlements
- Implemented comprehensive error handling and validation

Tests:
- Added comprehensive test suite for XvPSettlement contract
- Implemented tests for various settlement scenarios
- Covered edge cases like partial approvals, cancellations, and expirations

Chores:
- Removed legacy DvPSwap contract and related files
- Updated subgraph schema and mappings
- Updated contract deployment modules

### CODE REVIEW

File: kit/contracts/contracts/DvPSwapFactory.sol [hunk 1]
```diff
@@ -15,61 +15,65 @@ contract DvPSwapFactory is ReentrancyGuard, ERC2771Context {
...
     /// @param creator The address that created the swap contract
-    event DvPSwapContractCreated(address indexed dvpSwapContract, address indexed creator);
+    event DvPSwapCreated(address indexed token, address indexed creator);
 
     /// @notice Deploys a new DvPSwapFactory contract
     /// @dev Sets up the factory with meta-transaction support
     /// @param forwarder The address of the trusted forwarder for meta-transactions
-    constructor(address forwarder) ERC2771Context(forwarder) { }
+    constructor(address forwarder) ERC2771Context(forwarder) {
+        if (forwarder == address(0)) revert ZeroAddressNotAllowed();
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwapFactory.sol:36]: think we can add this `ZeroAddressNotAllowed` to all factories?

File: kit/contracts/contracts/DvPSwapFactory.sol [hunk 2]
```diff
@@ -15,61 +15,65 @@ contract DvPSwapFactory is ReentrancyGuard, ERC2771Context {
     /// @notice Custom errors for the DvPSwapFactory contract
     /// @dev These errors provide more gas-efficient and descriptive error handling
     error AddressAlreadyDeployed();
+    error ZeroAddressNotAllowed();
 
     /// @notice Mapping to track if an address was deployed by this factory
     /// @dev Maps swap contract addresses to a boolean indicating if they were created by this factory
-    mapping(address => bool) public isDvPSwapFromFactory;
+    mapping(address => bool) public isFactoryToken;
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwapFactory.sol:22]: can we rename this to `isFactorySwap`, since this isn't a token?

File: kit/contracts/contracts/DvPSwapFactory.sol [hunk 3]
```diff
@@ -15,61 +15,65 @@ contract DvPSwapFactory is ReentrancyGuard, ERC2771Context {
...
+
+    /// @notice Array of all swap contracts created by this factory
+    address[] public allSwaps;
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwapFactory.sol:25]: think we should remove this array, and use the map in the tests, it is not used anywhere but the tests so it increases the storage unnecessarily

File: kit/contracts/contracts/DvPSwapFactory.sol (hunk 2 above)
Comment (snigdha920) [reply kit/contracts/contracts/DvPSwapFactory.sol:22]: also the other spots where we use `token` can be renamed to `swap`

File: kit/contracts/ignition/modules/dvpswap.ts [hunk 4]
```diff
@@ -1,44 +1,43 @@
 // SPDX-License-Identifier: MIT
 // SettleMint.com
 import { buildModule } from "@nomicfoundation/hardhat-ignition/modules";
-// Import ethers correctly for keccak256
+// Remove unused ethers utils import
```
Comment (snigdha920) [inline kit/contracts/ignition/modules/dvpswap.ts:4]: think we can remove these comments

File: kit/contracts/contracts/DvPSwap.sol [hunk 5]
```diff
@@ -8,137 +8,166 @@ import { Pausable } from "@openzeppelin/contracts/utils/Pausable.sol";
...
-    error TimelockedFunds();
     error MaxDurationExceeded();
+    error AlreadyApproved();
+    error NotApproved();
+    error NotInvolved();
+    error DvPSwapAlreadyExecuted();
+    error DvPSwapNotApproved();
+    error InvalidToken();
     error FailedExecution(string reason);
+    error InsufficientAllowance(address token, address owner, address spender, uint256 required, uint256 allowed);
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:77]: i think this error should be thrown from the erc20 and we shouldn't duplicate it here

File: kit/contracts/contracts/DvPSwap.sol [hunk 6]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+            if (flow.amount == 0) revert ZeroAmount();
+            
+            // Validate ERC20 token by checking if it has decimals() function
+            (bool success, bytes memory result) = flow.token.staticcall(
+                abi.encodeWithSelector(bytes4(0x313ce567)) // decimals()
+            );
+            if (!success || result.length != 32) revert InvalidToken();
+            
+            dvpSwap.flows.push(flow);
+            emit FlowAdded(dvpSwapId, flow.from, flow.to, flow.token, flow.amount);
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:266]: i don't think we need a flow added event, just DvP swap created is enough

File: kit/contracts/contracts/DvPSwap.sol [hunk 7]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
-            tokenToSend,
-            tokenToReceive,
-            amountToSend,
-            amountToReceive,
-            timelock,
-            hashlock
-        );
+        // Mark swap as exists and set status to OPEN
+        dvpSwapExists[dvpSwapId] = true;
+        dvpSwap.status = DvPSwapStatus.OPEN;
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:271]: we already set the status to OPEN before the for loop

File: kit/contracts/contracts/DvPSwap.sol [hunk 8]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+        //         isInvolved = true;
+        //         break;
+        //     }
+        // }
+        // if (!isInvolved) revert NotInvolved();
+        
+        // Check allowance for all flows initiated by the caller
+        uint256 flowsLength = dvpSwap.flows.length;
+        for (uint256 i = 0; i < flowsLength; i++) {
+            Flow storage flow = dvpSwap.flows[i];
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:322]: think we can first take the sum of all the tokens that the caller will need to transfer, and then check once if the contract's allowance >= sum, instead of checking for every individual value

File: kit/contracts/contracts/DvPSwap.sol [hunk 9]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+        }
         
         // Update swap status
-        swap.status = DvPSwapStatus.REFUNDED;
-        
-        emit DvPSwapRefunded(dvpSwapId, swap.sender);
-        emit DvPSwapStatusChanged(dvpSwapId, DvPSwapStatus.REFUNDED);
+        dvpSwap.status = DvPSwapStatus.CLAIMED;
+        emit DvPSwapExecuted(dvpSwapId, _msgSender());
+        emit DvPSwapStatusChanged(dvpSwapId, DvPSwapStatus.CLAIMED);
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:412]: Don't think we need this 'status changed' event

File: kit/contracts/contracts/DvPSwap.sol [hunk 10]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+    // =========================================================================
+
+    /// @dev Modifier to check if a DvP swap exists.
+    modifier checkDvPSwapExists(bytes32 dvpSwapId) {
+        if (!dvpSwapExists[dvpSwapId]) revert DvPSwapNotFound();
+        _;
+    }
+
+    /// @dev Modifier to check if a DvP swap has not expired.
+    modifier checkDvPSwapNotExpired(bytes32 dvpSwapId) {
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:639]: since we check if a swap is expired or not solely based on block timestamp, i think we should remove the `EXPIRED` as a separate status on the swap. so i would remove the markDvPSwap failed method too since we don't need it

File: kit/contracts/contracts/DvPSwap.sol [hunk 11]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+
+    /// @dev Modifier to check if a DvP swap has expired.
+    modifier checkDvPSwapIsExpired(bytes32 dvpSwapId) {
+        DvPSwapData storage dvpSwap = _dvpSwaps[dvpSwapId];
+        if (block.timestamp <= dvpSwap.cutoffDate) revert InvalidDvPSwapParameters(); // Reusing error for "not expired yet"
+        _;
+    }
+
+    /// @dev Modifier to check if swap status is OPEN (suitable for actions like approve, execute, revoke, expire).
+    modifier checkStatusIsOpenForAction(bytes32 dvpSwapId) {
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:653]: how about just `onlyOpen` as the name?

Comment (snigdha920) [reply kit/contracts/contracts/DvPSwap.sol:653]: so it matches more with the openzepellin contract modifiers

File: kit/contracts/contracts/DvPSwap.sol [hunk 12]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+    modifier checkStatusIsOpenForCancel(bytes32 dvpSwapId) {
+        DvPSwapData storage dvpSwap = _dvpSwaps[dvpSwapId];
+        // Only check for OPEN status
+        if (dvpSwap.status != DvPSwapStatus.OPEN) 
+            revert InvalidDvPSwapStatus();
+        _;
+    }
+
+    /// @dev Modifier to check if swap status allows marking as failed.
+    modifier checkStatusAllowsFailure(bytes32 dvpSwapId) {
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:671]: this one can go too, since we don't want to allow marking a swap as 'failed', feels like the kind of state that could introduce loopholes too

File: kit/contracts/contracts/DvPSwap.sol [hunk 13]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+                isInvolved = true;
+                break;
+            }
+        }
+        if (!isInvolved) revert NotInvolved();
+        _;
+    }
+
+    /// @dev Modifier to check if the caller is the creator of the swap.
+    modifier checkCallerIsCreator(bytes32 dvpSwapId) {
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:726]: `onlyCreator`?

File: kit/contracts/contracts/DvPSwap.sol [hunk 14]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+
+    /// @dev Modifier to check if the caller has already approved the swap.
+    modifier checkCallerIsApproved(bytes32 dvpSwapId) {
+        DvPSwapData storage dvpSwap = _dvpSwaps[dvpSwapId];
+        if (!dvpSwap.approvals[_msgSender()]) revert NotApproved();
+        _;
+    }
+
+    /// @dev Modifier to check if the caller is a sender in any flow.
+    modifier checkCallerIsInvolvedSender(bytes32 dvpSwapId) {
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:697]: `onlySender`?

File: kit/contracts/contracts/DvPSwap.sol [hunk 15]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
-        
-        emit DvPSwapStatusChanged(dvpSwapId, DvPSwapStatus.INVALID);
-        
-        return true;
+        if (!isInvolved) revert NotInvolved();
+        _;
+    }
+
+    /// @dev Modifier to check if the caller is involved (sender or receiver) in any flow.
+    modifier checkCallerIsInvolvedAny(bytes32 dvpSwapId) {
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:711]: `onlyInvolved`?

File: kit/contracts/contracts/DvPSwap.sol [hunk 16]
```diff
@@ -163,453 +192,564 @@ contract DvPSwap is ReentrancyGuard, Pausable, AccessControl, ERC2771Context, Ow
...
+        }
+        if (!isAuthorized && hasRole(DEFAULT_ADMIN_ROLE, _msgSender())) {
+            isAuthorized = true;
+        }
+        if (!isAuthorized) revert NotAuthorized();
+        _;
+    }
+
+    /// @dev Modifier to check if the swap is fully approved by all senders.
+    modifier checkFullyApproved(bytes32 dvpSwapId) {
```
Comment (snigdha920) [inline kit/contracts/contracts/DvPSwap.sol:751]: `onlyFullyApproved`?

**Review by roderik — APPROVED** [review]
